### PR TITLE
⚠️ Add go modules to IPAM API

### PR DESCRIPTION
**Pull request by furkatgofurov7**
**What this PR does / why we need it**:
This PR adds go modules to IPAM API. The reason to add the go modules in API is to uplift IPAM for CAPM3. List of modules updated:

- sigs.k8s.io/controller runtime from `v0.7.0` to `v0.9.0-beta.0`
- sigs.k8s.io/cluster-api from `v0.3.12` to `v0.3.11-0.20210429135050-9d825629d89c` [v1alpha4](https://pkg.go.dev/sigs.k8s.io/cluster-api@v0.3.11-0.20210429135050-9d825629d89c/api/v1alpha4)
- k8s.io/client-go from `v1.5.1` to `v0.21.0`
- k8s.io/apimachinery from `v0.20.1` to `v0.21.0`
- k8s.io/apiextensions-apiserver from `v0.20.1` to `v0.21.0`
- k8s.io/api from `v0.20.1` to `v0.21.0`

**Important:**
Once this PR goes in, those who need to uplift IPAM, and they are only using IPAM APIs , you need to change your go modules to use

```
github.com/metal3-io/ip-address-manager/api/v1alpha1
```
 instead of:
```
github.com/metal3-io/ip-address-manager
```

If you are using/referring to other parts of code in IPAM not residing in API/v1alpha1, you need to use both the modules:
```
github.com/metal3-io/ip-address-manager/api/v1alpha1
github.com/metal3-io/ip-address-manager
```

### CODE REVIEW

**Comment by kashifest** [comment]
This looks good to me. Do we need to run integrations test?

**Comment by furkatgofurov7** [comment]
> This looks good to me. Do we need to run integrations test?

Most probably. 

/test-integration

**Comment by kashifest** [comment]
/approve
I believe this wont pass test-integration with the current changes right ?

File: ipam/utils.go [hunk 1]
```diff
@@ -56,24 +55,24 @@ func (e *NotFoundError) Error() string {
 	return "Object not found"
 }
 
-func updateObject(cl client.Client, ctx context.Context, obj runtime.Object, opts ...client.UpdateOption) error {
-	err := cl.Update(ctx, obj.DeepCopyObject(), opts...)
+func updateObject(cl client.Client, ctx context.Context, obj client.Object, opts ...client.UpdateOption) error {
+	err := cl.Update(ctx, obj.DeepCopyObject().(client.Object), opts...)
```
Comment (maelk) [inline ipam/utils.go:59]: casting an interface like this normally return 2 values, and you should check the second one to verify the casting happened properly.

**Review by maelk — APPROVED** [review]

**Comment by furkatgofurov7** [comment]
/close

go modules in this patch located under `/api/v1alpha1` which is not ideal as there might be a `/api/v1alpha2` in future and we need to consolidate modules outside of the `v1alphaX` folder anyways. Moving them out to `/api` in the current patch created some issues. I opened another [PR](https://github.com/metal3-io/ip-address-manager/pull/32) to have go.mod under `/api`.

**Pull request closed without merging**